src/Sparrow.Server/Collections/SmallSet.cs

```diff
@@ -0,0 +1,281 @@
+                _overflowStorage[key] = value;
+
+            _keys[idx] = key;
+            _values[idx] = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindKey(TKey key)
+        {
+            int elementIdx = Math.Min(_currentIdx, _length - 1);
+
+            var keys = _keys;
+            if (Vector.IsHardwareAccelerated && elementIdx > Vector256<TKey>.Count)
+            {
+                Vector<TKey> chunk;
+                var keyVector = new Vector<TKey>(key);
+                while (elementIdx >= Vector256<TKey>.Count)
+                {
+                    // We subtract because we are going to use that even in the case when there are differences.
+                    elementIdx -= Vector256<TKey>.Count;
+                    chunk = new Vector<TKey>(keys, elementIdx + 1);
+                    chunk = Vector.Equals(keyVector, chunk);
+                    if (chunk == Vector<TKey>.Zero)
+                        continue;
+
+                    elementIdx = Math.Min(elementIdx + Vector256<TKey>.Count, _length - 1);
+                    goto Found;
+                }
+
+                chunk = new Vector<TKey>(keys);
+                chunk = Vector.Equals(keyVector, chunk);
+                if (chunk == Vector<TKey>.Zero)
+                    return Invalid;
+
+                elementIdx = Vector256<TKey>.Count - 1;
+
+            Found:
+                elementIdx = Math.Min(elementIdx, _currentIdx);
+            }
+
+            while (elementIdx >= 0)
+            {
+                ref TKey current = ref keys[elementIdx];
+                if (current.Equals(key))
+                    return elementIdx;
+
+                elementIdx--;
+            }
+
+            return Invalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int RequestWritableBucket()
+        {
+            if (_currentIdx >= _length && _overflowStorage == null)
+            {
+                var storage = new Dictionary<TKey, TValue>(_currentIdx * 2);
+                for (int i = 0; i <= _currentIdx; i++)
+                    storage[_keys[i]] = _values[i];
+                _overflowStorage = storage;
+            }
+            
+            _currentIdx++;
+            return _currentIdx % _length;
+        }
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            // PERF: We put this into another method call to shrink the size of TryGetValue in the cases
+            // where the inliner would decide to inline the method. Given this method will be rarely executed
+            // as if it happens, probably this data structure is not the correct answer; the inliner will 
+            // most likely not inline this method ever. 
+            bool TryGetValueFromOverflowUnlikely(out TValue value)
+            {
+                if (_overflowStorage.TryGetValue(key, out value))
+                {
+                    _currentIdx++;
```

Comment:
We do _not_ need to worry about concurrent usage here, right?

This is on top of a dictionary, but it is mutable on reads.

Reply:
It is not synchronized if that is what you mean. not intended to be used in concurrent scenarios without locking (the same as dictionary).

Reply:
Let me see if I understand, you are worried about mutability of the cache (the LRU algorithm) where multiple readers can read. This is not a threaded aware data structure. NO, it is not a valid data structure for threaded scenarios. We could build a threaded scenario using `Interlocked` operations at the expense of locking operations. I didn't designed it to be used under that scenario but if you think that it makes sense just in case we mess up somewhere, either we can marked this in the name somehow OR perform this operation under a write barrier paying the performance cost.

Reply:
I don't think we want to use `Interlocked`.
My concern is that this is a mutable under read conditions.
I would rather *remove* the code that copy from the overflow storage to the set.

I assume that this is meant to be very rare, and it is better to have the additional cost in a rare scenario rather than have a potential issue down the line.

Reply:
Fair enough. As again it should be a very very rare occurence. More likely a edge case, if not an outright mis-sizing on our part.

Reply:
Yes, and having proper failure mode is best here.
Agree that should never happen